Add support for map type

yordis opened the pull request:
## Context

This is adding support for `map` based on https://github.com/WebAssembly/component-model/pull/554

## References

- https://github.com/bytecodealliance/wasm-tools/pull/2356
- https://github.com/WebAssembly/component-model/pull/554

yordis commented on .gitignore:
```diff
@@ -22,6 +22,7 @@ publish
 vendor
 examples/build
 examples/.cache
+crates/c-api/build
```
Is it ok to do this? I kept getting things to commit locally, so I added this

alexcrichton commented on crates/test-util/src/component.rs:
```diff
@@ -54,6 +54,7 @@ pub fn config() -> Config {
 
     let mut config = Config::new();
     config.wasm_component_model(true);
+    config.wasm_features(wasmtime::WasmFeatures::CM_MAP, true);
```
Could this get scoped to just the tests in question using this feature?

alexcrichton commented on crates/test-util/src/component_fuzz.rs:
```diff
@@ -158,6 +159,9 @@ impl Type {
         *fuel = fuel.saturating_sub(1);
+        // Note: Map types (case 21) are disabled because HashMap doesn't implement
+        // ComponentType, Lift, or Lower yet. When those implementations are added,
+        // change this to 21.
```
Is this still applicable?

yordis commented:
@alexcrichton addressed the comments, about the sharing stuff, feedback welcome there, I am not sure if it is the best way or not; not sure what you had in mind

alexcrichton commented on crates/wasmtime/src/runtime/component/values.rs:
```diff
@@ -939,6 +975,57 @@ fn load_list(cx: &mut LiftContext<'_>, ty: TypeListIndex, ptr: usize, len: usize
+    let tuple_alignment = key_abi.align32.max(value_abi.align32);
+    let tuple_size = usize::try_from(offset).unwrap();
+    let tuple_size = (tuple_size + usize::try_from(tuple_alignment)? - 1)
+        & !(usize::try_from(tuple_alignment)? - 1);
```
That might also be a good location to store a `value_offset: u32` value which is the offset from the start of the tuple to the value, so it also wouldn't need to be recomputed here for example.
